text/2018-10-25-tikv-client-rust.md

```diff
@@ -0,0 +1,389 @@
+
+        raw.get(&key)
+            .cf("test_cf")
+            .wait()
+            .expect_err("Get returned value for not existing key");
+
+        let keys = vec![b"k1".to_vec().into(), b"k2".to_vec().into()];
+
+        let _values = raw
+            .batch_get(&keys)
+            .cf("test_cf")
+            .wait()
+            .expect("Could not get values");
+
+        let start: Key = b"k1".to_vec().into();
+        let end: Key = b"k2".to_vec().into();
+        raw.scan(&start..&end, 10)
+            .cf("test_cf")
+            .key_only()
+            .wait()
+            .expect("Could not scan");
+
+        let ranges = [&start..&end, &start..&end];
+        raw.batch_scan(&ranges, 10)
+            .cf("test_cf")
+            .key_only()
+            .wait()
+            .expect("Could not batch scan");
+    }
+    ```
+
+The result is like:
+
+    ```bash
+    Successfully put Key([67, 111, 109, 112, 97, 110, 121]):Value([80, 105, 110, 103, 67, 65, 80]) to tikv
+    Found val: Value([80, 105, 110, 103, 67, 65, 80]) for key: Key([67, 111, 109, 112, 97, 110, 121])
+    Key: Key([67, 111, 109, 112, 97, 110, 121]) deleted
+    ```
+
+Raw Key-Value client is a client of the TiKV server and only supports the GET/BATCH_GET/PUT/BATCH_PUT/DELETE/BATCH_DELETE/SCAN/BATCH_SCAN/DELETE_RANGE commands. The Raw Key-Value client can be safely and concurrently accessed by multiple threads. Therefore, for one process, one client is enough generally.
+
+### Try the Transactional Key-Value API
+
+The Transactional Key-Value API is more complicated than the Raw Key-Value API. Some transaction related concepts are listed as follows.
+
+- Client
+
+    Like the Raw Key-Value client, a Client is client to a TiKV cluster.
+
+- Snapshot
+
+    A Snapshot is the state of a Client at a particular point of time, which provides some readonly methods. The multiple reads of the same Snapshot is guaranteed consistent.
+
+- Transaction
+
+    Like the Transaction in SQL, a Transaction symbolizes a series of read and write operations performed within the Client. Internally, a Transaction consists of a Snapshot for reads, and a buffer for all writes. The default isolation level of a Transaction is Snapshot Isolation.
+
+To use the Transactional Key-Value API, take the following steps:
+
+1. Create an instance of Config to specify endpoints of PD (Placement Driver) and optional security config
+
+    ```rust
+    let config = Config::new(vec!["127.0.0.1:2379"]);
+    ```
+
+2. Create a Transactional Key-Value client.
+
+    ```rust
+    let client = TxnClient::new(&config);
+    ```
+
+3. (Optional) Modify data using a Transaction.
+
+    The lifecycle of a Transaction is: _begin → {get, set, delete, scan} → {commit, rollback}_.
+
+4. Call the Transactional Key-Value API's methods to access the data on TiKV. The Transactional Key-Value API contains the following methods:
+
+    ```rust
+    fn begin(&self) -> KvFuture<Transaction>;
```

Comment:
What is KvFuture?

Reply:
It's added to common data types section now. Thanks